tracing/hooks.go

```diff
@@ -3,17 +3,31 @@ package tracing
 // BaseplateHook allows you to inject functionality into the lifecycle of a
 // Baseplate request.
 type BaseplateHook interface {
-	OnServerSpanCreate(span *Span) error
+	OnServerSpanCreate(span ServerSpan) error
 }
 
 // SpanHook allows you to inject functionality into the lifecycle of Baseplate
 // Spans.
 type SpanHook interface {
-	OnCreateChild(child *Span) error
-	OnStart(span *Span) error
-	OnEnd(span *Span, err error) error
-	OnSetTag(span *Span, key string, value interface{}) error
-	OnAddCounter(span *Span, key string, delta float64) error
+	OnCreateChild(child Span) error
+	OnStart(span Span) error
+	OnEnd(span Span, err error) error
+	OnSetTag(span Span, key string, value interface{}) error
+	OnAddCounter(span Span, key string, delta float64) error
+}
+
+type SpanHooks struct {
+	hooks []SpanHook
+}
+
+// AddHooks adds hooks into the SpanHooks container.
+func (h *SpanHooks) AddHooks(hooks ...SpanHook) {
+	h.hooks = append(h.hooks, hooks...)
+}
+
+// Hooks returns a copy of the internal list of SpanHooks.
+func (h *SpanHooks) Hooks() []SpanHook {
+	return h.hooks
```

Comment:
This is not actually a copy. For example the follow code could change your 2nd hook:

```
foo := hooks.Hooks()
foo[1] = newHook
```

Reply:
I wonder if it's useful to define `SpanHooks` type, or maybe just typedef it as `type SpanHooks []SpanHook`.